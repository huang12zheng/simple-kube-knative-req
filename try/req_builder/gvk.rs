use kube::api::GroupVersionKind;

pub struct GKVSpec {
    pub group: String,
    pub version: String,
    pub kind: String,
}

impl GKVSpec {
    #[allow(dead_code)]
    pub fn from_env() -> Self {
        let group = std::env::var("GROUP").unwrap_or_else(|_| "serving.knative.dev".into());
        let version = std::env::var("VERSION").unwrap_or_else(|_| "v1".into());
        let kind = std::env::var("KIND").unwrap_or_else(|_| "Service".into());
        Self {
            group,
            version,
            kind,
        }
    }
}

impl Into<GroupVersionKind> for GKVSpec {
    fn into(self) -> GroupVersionKind {
        let gkv = GroupVersionKind::gvk(&self.group, &self.version, &self.kind);
        gkv
    }
}
