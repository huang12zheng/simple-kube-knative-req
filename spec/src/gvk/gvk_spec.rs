use crate::*;
/// GVKSpec is spec of [GroupVersionKind]
/// # Example For Default Data
/// ```snap
#[doc = include_str!("snapshots/kube_do_spec__gvk__gvk_spec__tests__gvk_spec.snap")]
/// ```
#[doc(alias = "gvk")] // for Support search function in doc
#[derive(Envconfig, Debug, Clone)]
pub struct GVKSpec {
    /// source from [GroupVersionKind::group]
    #[envconfig(from = "GROUP", default = "serving.knative.dev")]
    pub group: String,
    /// source from [GroupVersionKind::version]
    #[envconfig(from = "VERSION", default = "v1")]
    pub version: String,
    /// source from [GroupVersionKind::kind]
    #[envconfig(from = "KIND", default = "Service")]
    pub kind: String,
}

/// From generic [T] into [GVKSpec], such as [crate::ksvc::KnativeSpec]
/// Detail are in [crate::ksvc::KnativeSpec::into_gvk]
pub trait IntoGVKSpec {
    /// From [KnativeSpec] into [GVKSpec]
    fn into_gvk(&self) -> GVKSpec;
}

impl From<GVKSpec> for GroupVersionKind {
    fn from(val: GVKSpec) -> Self {
        let gkv = GroupVersionKind::gvk(&val.group, &val.version, &val.kind);
        gkv
    }
}

#[cfg(test)]
mod tests {
    use crate::gvk::GVKSpec;
    use envconfig::Envconfig;

    /// [GVKSpec::init_from_env]
    /// snap would use by doc of [GVKSpec]
    #[test]
    fn gvk_spec() {
        let gvk = GVKSpec::init_from_env().unwrap();
        insta::assert_debug_snapshot!(gvk);
    }
}
