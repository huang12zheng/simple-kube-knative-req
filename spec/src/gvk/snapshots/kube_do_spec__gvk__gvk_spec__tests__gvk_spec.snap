---
source: spec/src/gvk/gvk_spec.rs
expression: gvk
---
GVKSpec {
    group: "serving.knative.dev",
    version: "v1",
    kind: "Service",
}
